core/modulesource.go

```diff
@@ -245,6 +249,150 @@ func (src *ModuleSource) AutomaticGitignore(ctx context.Context) (*bool, error)
 	return modCfg.Codegen.AutomaticGitignore, nil
 }
 
+// LoadContext loads a directory from the module context directory.
+//
+// If the module is local, it will load the directory from the local source
+// directly from the host.
+//
+// If the module is git, it will load the directory from the git repository
+// using its context directory.
+//
+// The path is relative to the module context directory, even if it's absolute.
+//
+// Improvement for later: use `.gitignore` to excludes paths from the context.
+func (src *ModuleSource) LoadContext(ctx context.Context, dag *dagql.Server, path string, ignore []string) (inst dagql.Instance[*Directory], err error) {
+	// We exclude .git by default because it's not useful and tends to invalidate the cache.
+	excludes := []string{"**/.git"}
+	includes := []string{}
+
+	for _, p := range ignore {
+		if strings.HasPrefix(p, "!") {
+			includes = append(includes, p[1:])
+		} else {
+			excludes = append(excludes, p)
+		}
+	}
+
+	// If path is not absolute, it's relative to the module root directory.
+	if !filepath.IsAbs(path) {
+		moduleRootPath, err := src.SourceRootSubpath()
+		if err != nil {
+			return inst, fmt.Errorf("failed to get source root subpath for relative context path: %w", err)
+		}
+
+		path = filepath.Join(moduleRootPath, path)
+	}
+
+	switch src.Kind {
+	case ModuleSourceKindLocal:
+		bk, err := src.Query.Buildkit(ctx)
+		if err != nil {
+			return inst, fmt.Errorf("failed to get buildkit api: %w", err)
+		}
+
+		slog.Debug("moduleSource.LoadContext: loading contextual directory from local source", "path", path, "kind", src.Kind)
```

Comment:
Just an FYI, you can print the type using the `%T` format specifier, instead of manually duplicating the information :tada: